Milestone 3: Game logic

Mupa1 opened the pull request:
## What does this PR do?
- It adds the game logic for Tic Tac Toe project

## Description of task completed
- Separated methods into different classes and files
- Added validations for user's names
- Refactored the play and won methods after TSC’s review
- Added the board class
- Added the user's turn to play after TSC’s review

Mupa1 committed:
separate methods into diffrent classes

Mupa1 committed:
refactor repeated code

okikiola11 committed:
Add validations

rOluochKe reviewed (changes requested):
Hello @Mupa1 and @okikiola11 
### You have done a good job so far :clap: However, you can do better before your project could be approved.
- [ ] You can create a game module and put it in the lib. This makes the main.rb file to only call the files and runs the loop making your code easily readable and maintainable
- [ ] When a player wins a game, it is supposed to call the winner's name, not `0`
- [ ] It is also good practice when the game ends, it asks if the players want to continue or quit the game

### Make all these changes and resubmit for another code review. Good luck.
- Happy coding!

By Raymond
slack @iamraymond

Mupa1 committed:
saparated helper methods into the game module

okikiola11 committed:
Add the restart game method to the main.rb file

okikiola11 committed:
Fix linting errors

Mupa1 committed:
update play method on main.rb file

pelzolga123 reviewed (changes requested):
The project requires changes.

pelzolga123 commented:
## Hi Team,
### Amazing work so far :+1:
Still, there are some more changes required before this Pull Request can be approved: 
* [ ] All classes should be kept in the `lib` directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)) You have the PlayGame class inside the `main` file, but all classes should be stored in the `lib` directory. 
* [ ] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). Don't create long methods that are doing multiple steps of the game process. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)) You have method `play` in your project, that is pretty big, that should be fixed.

### Please kindly fix the above issues and send another code review request.

If you have any questions or need help, you can message me on slack (@ Olga Pelts). *Happy coding*

Mupa1 committed:
refactor code

Mupa1 committed:
fix lint errors

majovanilla commented on bin/main.rb:
```diff
@@ -1,107 +1,68 @@
-      counter += 1 if spaces.include?(@player1) || spaces.include?(@player2)
-    end
-    counter
+def welcome
+  puts 'Welcome to Tic Tac Toe'
+  puts 'Please Enter your name (Player 1): '
+  player1_name = gets.strip.upcase
+  until @playerz.validate_name(player1_name)
+    puts 'Player 1: Enter Valid name'
+    player1_name = gets.chomp.upcase
   end
```
Great validation! :+1:  you should add a check to prevent both players to have the same name.

majovanilla commented on bin/main.rb:
```diff
@@ -1,107 +1,68 @@
-
-  def current_player
-    turn_count(@board).even? ? @player1 : @player2
+  puts 'Please Enter your name (Player 2): '
+  player2_name = gets.strip.upcase
+  until @playerz.validate_name(player2_name)
+    puts 'Player 2: Enter Valid name'
+    player2_name = gets.chomp.upcase
   end
```
You are repeating this code for each player. That is not a good practice (see  this article's [Principle 3](https://www.nascenia.com/ruby-on-rails-development-principles/))
You could easily move everything into another method and call it two times.
Remember to avoid `puts` on the `lib` folder when you refactor this code. You can still return a string and then `puts` it on `main.rb`.

majovanilla commented on bin/main.rb:
```diff
@@ -1,107 +1,68 @@
+  welcome
+  game_over = false
+  puts @game.display_board
+  until game_over == true
+    puts 'Select a number between 1-9'
+    input = @game.input_to_index(gets.chomp)
```
You don't show which player's turn is. This causes confusion and affects the UX.

majovanilla commented on bin/main.rb:
```diff
@@ -1,107 +1,68 @@
+  until game_over == true
+    puts 'Select a number between 1-9'
+    input = @game.input_to_index(gets.chomp)
+    if !@game.valid_move?(input)
+      puts 'Invalid move, number already taken or out of range'
+    else
+      @game.move(input, @game.current_player)
+      if @game.won?
+        if @game.current_player == 'O'
+          puts "#{@playerz.player1} won the game!"
+        else
+          puts "#{@playerz.player2} won the game!"
         end
+        puts @game.display_board
+        return game_over == true
+      elsif !@game.board.include?(' ')
+        puts @game.display_board
+        puts 'It\'s a Draw!'
+        return game_over == true
```
You have three nested if statements nested in a loop in this method. I strongly suggest people avoid this since it's not efficient and it's difficult to read. If you must use them, I would advise you to create separate methods for each if statement, improving readability.

majovanilla commented on bin/main.rb:
```diff
@@ -1,107 +1,68 @@
-      true
-    end
-  end
+def play_again
+  game_over = true
+  puts 'Would you like to restart the game (Yes/No)'
+  player_response = gets.chomp.downcase
+  return game_over if player_response != 'yes'
+
+  @game.empty_board
+  play
+  play_again
```
Your method breaks when the user types something else than yes or no. You should validate this and ask for the response again to prevent mistakes.

majovanilla commented on lib/game.rb:
```diff
@@ -0,0 +1,67 @@
+  end
+
+  # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+  def won?
+    if (@board[0] == @board[1] && @board[1] == @board[2]) && !@board.values_at(0, 1, 2).include?(' ')
+      true
+    elsif (@board[3] == @board[4] && @board[4] == @board[5]) && !@board.values_at(3, 4, 5).include?(' ')
+      true
+    elsif (@board[6] == @board[7] && @board[7] == @board[8]) && !@board.values_at(6, 7, 8).include?(' ')
+      true
+    elsif (@board[0] == @board[3] && @board[3] == @board[6]) && !@board.values_at(0, 3, 6).include?(' ')
+      true
+    elsif (@board[1] == @board[4] && @board[4] == @board[7]) && !@board.values_at(1, 4, 7).include?(' ')
+      true
+    elsif (@board[2] == @board[5] && @board[5] == @board[8]) && !@board.values_at(2, 5, 8).include?(' ')
+      true
+    elsif (@board[0] == @board[4] && @board[4] == @board[8]) && !@board.values_at(0, 4, 8).include?(' ')
+      true
+    elsif (@board[2] == @board[4] && @board[4] == @board[6]) && !@board.values_at(2, 4, 6).include?(' ')
+      true
+    end
+  end
+  # rubocop:enable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
```
You could improve this method substantially. TIP: try creating an array with all the possible winning combinations and then comparing it to the current board array to see if it `includes` any of them. Use the player's symbol as an argument to the method.

majovanilla commented on lib/game.rb:
```diff
@@ -0,0 +1,67 @@
+
+  attr_reader :board
+
+  def display_board
+    [
+      [" #{@board[0]} | #{@board[1]} | #{@board[2]} "],
+      ['----------- '],
+      [" #{@board[3]} | #{@board[4]} | #{@board[5]} "],
+      ['----------- '],
+      [" #{@board[6]} | #{@board[7]} | #{@board[8]} "]
+    ]
+  end
```
The Game class should have the game logic. It would be better if you create another class to have all the Board related methods, keeping your code more organized.

majovanilla reviewed (changes requested):
Hi team!

## You did a *great* job on this project! 💜 

There are still some changes 👇 you need to implement to approve this request 🔍
Check the commented code below.

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

Mupa1 committed:
update welcome method

okikiola11 committed:
Fixed code review
Add the board class
Update the play_again not to accept any argument outside yes or no

okikiola11 committed:
Fix linting errors

okikiola11 committed:
Update #play_again method on the main.rb file

Mupa1 committed:
refactor play method into smaller methods

Mupa1 committed:
refactor won method to win_combinations array

bolah2009 commented on lib/game.rb:
```diff
@@ -0,0 +1,93 @@
+  end
+
+  def full_board?
+    true unless @board.board_cells.include?(' ')
```
[Optional]

I think we could use only `!@board.board_cells.include?(' ')` in this case. Or use methods we learn from the `Enumerable Projects`, a good example is the `none?` since this will return true if **none** of the `board_cells` is ` `. e.g. `@board.board_cells.none?(' ')` 🤔

bolah2009 commented on lib/game.rb:
```diff
@@ -0,0 +1,93 @@
+  end
+
+  def game_over?
+    true if full_board? || won?
```
[Optional]

Same with the suggestion above, I think we can return `full_board? || won?`, assuming `won?` will always return a boolean.

bolah2009 reviewed (approved):
It's a 👍🏾 from me!

### :heavy_check_mark: Approved!

Check out the optional comments before merging. :relaxed: 

> ### Let's remember about `private` methods, methods that are not used outside of the class should be private methods.

Good luck with your next projects. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

Mupa1 commented:
Thank you so much for your reviews, especially @majovanilla for your descriptive feedback.

okikiola11 committed:
Fix code review changes
